rfc/instance.md

```diff
@@ -0,0 +1,118 @@
+
+This API will provide a way to create one engine per instance. User can take advantage of this and create multiple spaniel instances, one with optimized engine and other with unoptimized engine, based on requirement.
+
+### Interface
+The spaniel instance interface looks something like below.
+
+```typescript
+
+interface SpanielIntersectionObserverInterface {
+  new(callback: Function, options: IntersectionObserverInit): SpanielIntersectionObserver;
+}
+
+interface SpanielObserverInterface {
+  new (callback: (entries: SpanielObserverEntry[]) => void, options: SpanielObserverInit): SpanielObserver;
+}
+
+interface WatcherInterface {
+  new(config: WatcherConfig): Watcher;
+}
+
+interface SpanielInstance {
+  public IntersectionObserver: IntersectionObserverClass;
+  private engine: Engine;
+  private scheduler: Scheduler;
+  public scheduleRead: (callback: Function) => void;
+  public scheduleWork: (callback: Function) => void;
+  public eventStore: EventStore;
+  public SpanielObserver: SpanielObserverInterface;
+  public Watcher: WatcherInterface;
+  getEventStore: () => EventStore;
+  on: (eventName: string, callback: (frame: FrameInterface, id: string) => void;
+  off: (eventName: string, callback: Function) => void;
+  trigger: (eventName: string, value?: any) => void;
+  getScheduler: () => Scheduler;
+  getEngine: () => Engine;
+}
+```
+- Each instance of spaniel will have its own `engine` and `scheduler`. These two properties are not supposed to be modified directly. So are making it private property of instance. To access the these two properties we will give two functions `getScheduler` and `getEngine`.
+- Each instance will have its own event store.
+- Each instance will have on, off, trigger functions to schedule/unschedule events in the instance.
+
+
+# How We Teach This
+
+Since we are allowing users to create new instances. User must changes exisiting direct usage of spaniel to creating new instances and then using it. 
+Example:
+
+Usage changes from
+```JavaScript
+import { IntersectionObserver } from 'spaniel';
+
+new IntersectionObserver((entries) => { console.log('I see you') }, {
+  threshold: 0.5
+}).observe(document.getElementById('my-element'));
+```
+to
+```JavaScript
+import { SpanielInstance } from 'spaniel';
+
+const spanielInstance = new SpanielInstance();
+new spanielInstance.IntersectionObserver((entries) => { console.log('I see you') }, {
+  threshold: 0.5
+}).observe(document.getElementById('my-element'));
+```
+
+All the correspoding functions which were used directly before must be used along with instantiated object. 
+Example 
+- `spaniel.on()` -> `spanielInstance.on()`
+- `spaniel.off()` -> `spanielInstance.off()`
+- `spaniel.trigger()` -> `spanielInstance.trigger()`
+
+# Drawback
+
+After this upgrade there is no way to use the spaniel without instantiating the spaniel. So user who use this upgraded version must update previous implementation. There will not be any backward compatibility.
```

Comment:
@rwjblue True, but the new API requires passing in `IntersectionObserver`. Going forward, we don't want to have `IntersectionObserver` in the module graph of `SpanielObserver/Instance/Watcher` to simplify setting up per-browser builds where we don't ship the `IntersectionObserver` polyfill to browsers that don't need it.